lmfit/parameter.py

```diff
@@ -38,15 +40,32 @@ class Parameters(OrderedDict):
     def __init__(self, asteval=None, *args, **kwds):
         super(Parameters, self).__init__(self)
         self._asteval = asteval
+
+        # a flag to temporarily pause updating constraints when lots of
+        # parameters are being added. This is mainly used during the deepcopy
+        # operation. Do not forget to switch back on.
+        self._dont_update_constraints = False
+
         if asteval is None:
             self._asteval = Interpreter()
         self.update(*args, **kwds)
 
     def __deepcopy__(self, memo):
         _pars = Parameters()
-        for key, val in self._asteval.symtable.items():
-            if key not in self._asteval.no_deepcopy:
-                _pars._asteval.symtable[key] = deepcopy(val, memo)
+
+        # find the symbols that were added by users, not during construction
+        sym_in_current = set(self._asteval.symtable.keys())
+        sym_from_construction = set(_pars._asteval.no_deepcopy)
+        sym_unique = sym_in_current.difference(sym_from_construction)
+        unique_symbols = {key: deepcopy(self._asteval.symtable[key], memo)
+                          for key in sym_unique}
+        _pars._asteval.symtable.update(unique_symbols)
+
+        # we're just about to add a lot of Parameter objects to the newly
+        # created object. Temporarily switch off updating_constraints().
+        # This causes a O(N**2) slowdown.
+        _pars._dont_update_constraints = True
```

Comment:
and if you are going to do it this way I would suggest giving `_pars` a context manager so that if something errors out inside, it will fall back to a safe state.